File: app/src/main/java/alexis/tvrs/rickandmortyuniverse/ui/characters/CharacterFragment.java
package alexis.tvrs.rickandmortyuniverse.ui.characters;

import android.content.Intent;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.GridView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;

import java.util.ArrayList;
import java.util.Objects;
import java.util.Random;

import alexis.tvrs.rickandmortyuniverse.NetworkUtil;
import alexis.tvrs.rickandmortyuniverse.R;
import alexis.tvrs.rickandmortyuniverse.SharedPreferencesCharacters;
import alexis.tvrs.rickandmortyuniverse.SharedPreferencesFavorites;

public class CharacterFragment extends Fragment {
    private static final int NBCHARACTERS = 493;
    private ArrayList<Character> characters = new ArrayList<>();
    private CharacterGridAdapter characterGridAdapter;
    private SharedPreferencesFavorites sharedPreferencesFavorites;
    private SharedPreferencesCharacters sharedPreferencesCharacters;
    private static CharacterFragment instance;

    public static CharacterFragment getInstance(){
        if(instance == null)
            instance = new CharacterFragment();

        return instance;
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View root = inflater.inflate(R.layout.fragment_characters, container, false);

        sharedPreferencesFavorites = new SharedPreferencesFavorites();
        sharedPreferencesCharacters = new SharedPreferencesCharacters();

        GridView simpleGrid = root.findViewById(R.id.fragment_characters_simpleGridView);
        characterGridAdapter = new CharacterGridAdapter(this.getContext(),characters);
        simpleGrid.setAdapter(characterGridAdapter);

        simpleGrid.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
                Intent intent = new Intent(getActivity(), CharacterFullActivity.class);
                intent.putExtra("CHARACTER",characters.get(position));
                startActivity(intent);
            }
        });

        simpleGrid.setOnItemLongClickListener(new AdapterView.OnItemLongClickListener() {
            @Override
            public boolean onItemLongClick(AdapterView<?> parent, View view, int position, long id) {
                sharedPreferencesFavorites.addFavorite(getActivity(),characters.get(position));
                //TODO Check if characters is already in favorites
                Toast.makeText(getActivity(), "Added to favorites", Toast.LENGTH_LONG).show();
                return true;
            }
        });


        final SwipeRefreshLayout pullToRefresh = root.findViewById(R.id.fragment_characters_pullToRefresh);
        pullToRefresh.setOnRefreshListener(new SwipeRefreshLayout.OnRefreshListener() {
            @Override
            public void onRefresh() {
                loadCharacters();
                pullToRefresh.setRefreshing(false);
            }
        });

        loadCharacters();
        return root;
    }

    private void loadCharacters(){
        if(NetworkUtil.getConnectivityStatus(Objects.requireNonNull(getActivity()))){
            loadOnline();
        }
        else {
            loadOffline();
        }
    }

    private void loadOffline(){
        Toast.makeText(getActivity(), R.string.networkError, Toast.LENGTH_SHORT).show();
        characters = (sharedPreferencesCharacters.getcharacters(Objects.requireNonNull(getActivity())));
        characterGridAdapter.notifyDataSetChanged();
    }

    private void loadOnline(){
        characters.clear();
        ArrayList<Integer> ids = new ArrayList<>(5);
        Random r = new Random();
        int randomNumber;
        for(int i = 0; i < 8; i++) {
            randomNumber =  r.nextInt(NBCHARACTERS) + 1;
            if(!ids.contains(randomNumber)) {
                ids.add(randomNumber);
                new AsyncDownloadCharacter().execute(randomNumber, characters, characterGridAdapter);
            }
        }
        sharedPreferencesCharacters.savecharacters(Objects.requireNonNull(getActivity()),characters);
    }
}

File: app/src/main/java/alexis/tvrs/rickandmortyuniverse/ui/characters/CharacterFragment.kt
package alexis.tvrs.rickandmortyuniverse.ui.characters

import alexis.tvrs.rickandmortyuniverse.NetworkUtil.getConnectivityStatus
import alexis.tvrs.rickandmortyuniverse.R
import alexis.tvrs.rickandmortyuniverse.SharedPreferencesCharacters
import alexis.tvrs.rickandmortyuniverse.SharedPreferencesFavorites
import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView.OnItemClickListener
import android.widget.AdapterView.OnItemLongClickListener
import android.widget.GridView
import android.widget.Toast
import androidx.fragment.app.Fragment
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import java.util.*

class CharacterFragment : Fragment() {
    private var characters: ArrayList<Character>? = ArrayList()
    private var characterGridAdapter: CharacterGridAdapter? = null
    private var sharedPreferencesFavorites: SharedPreferencesFavorites? = null
    private var sharedPreferencesCharacters: SharedPreferencesCharacters? = null
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val root = inflater.inflate(R.layout.fragment_characters, container, false)
        sharedPreferencesFavorites = SharedPreferencesFavorites()
        sharedPreferencesCharacters = SharedPreferencesCharacters()
        val simpleGrid = root.findViewById<GridView>(R.id.fragment_characters_simpleGridView)
        characterGridAdapter = CharacterGridAdapter(this.context, characters)
        simpleGrid.adapter = characterGridAdapter
        simpleGrid.onItemClickListener = OnItemClickListener { parent, view, position, id ->
            val intent = Intent(activity, CharacterFullActivity::class.java)
            intent.putExtra("CHARACTER", characters!![position])
            startActivity(intent)
        }
        simpleGrid.onItemLongClickListener = OnItemLongClickListener { parent, view, position, id ->
            sharedPreferencesFavorites!!.addFavorite(requireActivity(), characters!![position])
            //TODO Check if characters is already in favorites
            Toast.makeText(activity, "Added to favorites", Toast.LENGTH_LONG).show()
            true
        }
        val pullToRefresh: SwipeRefreshLayout = root.findViewById(R.id.fragment_characters_pullToRefresh)
        pullToRefresh.setOnRefreshListener {
            loadCharacters()
            pullToRefresh.isRefreshing = false
        }
        loadCharacters()
        return root
    }

    private fun loadCharacters() {
        if (getConnectivityStatus(requireActivity())) {
            loadOnline()
        } else {
            loadOffline()
        }
    }

    private fun loadOffline() {
        Toast.makeText(activity, R.string.networkError, Toast.LENGTH_SHORT).show()
        characters = sharedPreferencesCharacters!!.getcharacters(requireActivity())
        characterGridAdapter!!.notifyDataSetChanged()
    }

    private fun loadOnline() {
        characters!!.clear()
        val ids = ArrayList<Int>(5)
        val r = Random()
        var randomNumber: Int
        for (i in 0..7) {
            randomNumber = r.nextInt(NBCHARACTERS) + 1
            if (!ids.contains(randomNumber)) {
                ids.add(randomNumber)
                AsyncDownloadCharacter().execute(randomNumber, characters, characterGridAdapter)
            }
        }
        sharedPreferencesCharacters!!.savecharacters(requireActivity(), characters)
    }

    companion object {
        private const val NBCHARACTERS = 493
        var instance: CharacterFragment? = null
            get() {
                if (field == null) field = CharacterFragment()
                return field
            }
            private set
    }
}